### ISSUE TITLE
Operator silently deletes resources of other clusters CRD's

### ISSUE THREAD

**Issue by rongo89** [citrix/citrix-k8s-ingress-controller#257]
**Describe the bug**
We have a setup where we're running multiple k8s cluster, each with its own operator but accessing the same netscaler. There is for instance an rewritePolicy CRD defined for cluster A, but cluster B has no rewritePolicy CRD. What happens: operator of cluster B unbinds policy from service of cluster A and deletes policy after a while. This seems to happen *after* restart of the operator pod in cluster B

Each operator has its own NS_APPS_NAME_PREFIX defined. All resources have distinct names.

**To Reproduce**
1. steps
2. Version of the Citrix Ingress Controller: 1.7.6
3. Version of MPX: NS12.1 56.22.nc
4. Environment variables:
`NS_IP:                <nsIP>                                                                                                                                                                 
NS_VIP:               <nsVIP>                                                                                                                                                                 
NS_USER:              USERNAME                                                                                                                                                                
NS_PASSWORD:          PASSWORD                                                                                                                                                                
EULA:                 true                                                                                                                                                                          
LOGLEVEL:             DEBUG                                                                                                                                                                       
NS_PROTOCOL:          HTTPS                                                                                                                                                                              NS_PORT:              443                                                                                                                                                                         NS_APPS_NAME_PREFIX:  <nsNamespace>`

**Expected behavior**
cluster A operator and cluster B operator can act independently from each other on the same netscaler, and doesnt delete each others resource. 

**Logs**
From the netscaler logs, it indicates with an clientIP coming from cluster B, that it deletes resources beloging to cluster A
```
Jun 11 21:45:56 <local0.info> <ns_ip> 06/11/2020:21:45:56 GMT <ns_name>0-PPE-1 : default API CMD_EXECUTED 3710016 0 :  User <user> - Remote_ip <clientIP> - Command "unbind lb vse
rver <nsNamespace>-netscaler-internal-443_namespace_443_<nsNamespace>-namespace-nginx-ingress-internal-controller_namespace_443_svc -policyName k8s_rwpolicy_crd_proxyprotocol0<nsNamespace>_0_namespace_138971
989 -type REQUEST -priority 1004" - Status "Success"
Jun 11 21:45:56 <local0.info> <ns_ip> 06/11/2020:21:45:56 GMT <ns_name>0-PPE-1 : default API CMD_EXECUTED 3710017 0 :  User <user> - Remote_ip <clientIP> - Command "unbind lb vse
rver <nsNamespace>-netscaler-internal-80_namespace_80_<nsNamespace>-namespace-nginx-ingress-internal-controller_namespace_80_svc -policyName k8s_rwpolicy_crd_proxyprotocol0<nsNamespace>_0_namespace_138971989
 -type REQUEST -priority 1004" - Status "Success"
Jun 11 21:45:56 <local0.info> <ns_ip> 06/11/2020:21:45:56 GMT <ns_name>0-PPE-1 : default API CMD_EXECUTED 3710018 0 :  User <user> - Remote_ip <clientIP> - Command "rm rewrite po
licy k8s_rwpolicy_crd_proxyprotocol0<nsNamespace>_0_namespace_138971989" - Status "Success"
Jun 11 21:45:56 <local0.info> <ns_ip> 06/11/2020:21:45:56 GMT <ns_name>0-PPE-1 : default API CMD_EXECUTED 3710019 0 :  User <user> - Remote_ip <clientIP> - Command "rm rewrite ac
tion k8s_rwaction_crd_proxyprotocol0<nsNamespace>_0_namespace_138971989" - Status "Success
```
Operator logs
```
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:10,865  - DEBUG - [rewritepolicy.py:sync_done:209] (MainThread) REWRITEPOLICY CRD CONFIGURATION BEGINS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,038  - DEBUG - [rewritepolicy.py:create_obsolete_list:662] (MainThread) Obsolete Item: rwpolicy LIST BEGINS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,038  - DEBUG - [rewritepolicy.py:create_obsolete_list:664] (MainThread) Obsolete list 0: k8s_rwpolicy_crd_proxyprotocol0<ns_namespace>_0_namespace_139420825
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,038  - DEBUG - [rewritepolicy.py:create_obsolete_list:665] (MainThread) Obsolete item: rwpolicy LIST ENDS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,140  - DEBUG - [nitrointerface.py:unbind_lb_vserver_rewrite_policy:2721] (MainThread) Clubconfig DELETE: lb <ns_namespace>-netscaler-internal-443_namespace_443_<ns_namespace>-namespace-nginx-ingress-internal-controller_namespace_443_svc unbinding with rewritepolicy k8s_rwpolicy_crd_proxyprotocol0<ns_namespace>_0_namespace_139420825
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,199  - DEBUG - [nitrointerface.py:unbind_lb_vserver_rewrite_policy:2721] (MainThread) Clubconfig DELETE: lb <ns_namespace>-netscaler-internal-80_namespace_80_<ns_namespace>-namespace-nginx-ingress-internal-controller_namespace_80_svc unbinding with rewritepolicy k8s_rwpolicy_crd_proxyprotocol0<ns_namespace>_0_namespace_139420825
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,253  - DEBUG - [rewritepolicy.py:create_obsolete_list:662] (MainThread) Obsolete Item: resppolicy LIST BEGINS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,253  - DEBUG - [rewritepolicy.py:create_obsolete_list:665] (MainThread) Obsolete item: resppolicy LIST ENDS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,295  - DEBUG - [rewritepolicy.py:create_obsolete_list:662] (MainThread) Obsolete Item: resppolicy LIST BEGINS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,295  - DEBUG - [rewritepolicy.py:create_obsolete_list:665] (MainThread) Obsolete item: resppolicy LIST ENDS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,329  - DEBUG - [rewritepolicy.py:create_obsolete_list:662] (MainThread) Obsolete Item: respaction LIST BEGINS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,329  - DEBUG - [rewritepolicy.py:create_obsolete_list:665] (MainThread) Obsolete item: respaction LIST ENDS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,365  - DEBUG - [rewritepolicy.py:create_obsolete_list:662] (MainThread) Obsolete Item: respaudmesaction LIST BEGINS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,365  - DEBUG - [rewritepolicy.py:create_obsolete_list:665] (MainThread) Obsolete item: respaudmesaction LIST ENDS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,569  - DEBUG - [rewritepolicy.py:create_obsolete_list:662] (MainThread) Obsolete Item: rwpolicy LIST BEGINS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,569  - DEBUG - [rewritepolicy.py:create_obsolete_list:664] (MainThread) Obsolete list 0: k8s_rwpolicy_crd_proxyprotocol0<ns_namespace>_0_namespace_139420825
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,570  - DEBUG - [rewritepolicy.py:create_obsolete_list:665] (MainThread) Obsolete item: rwpolicy LIST ENDS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,601  - DEBUG - [nitrointerface.py:delete_ns_config:4297] (MainThread) delete policy rewrite policy k8s_rwpolicy_crd_proxyprotocol0<ns_namespace>_0_namespace_139420825 added to bulk config
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,846  - DEBUG - [rewritepolicy.py:create_obsolete_list:662] (MainThread) Obsolete Item: rwaction LIST BEGINS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,847  - DEBUG - [rewritepolicy.py:create_obsolete_list:664] (MainThread) Obsolete list 0: k8s_rwaction_crd_proxyprotocol0<ns_namespace>_0_namespace_139420825
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,847  - DEBUG - [rewritepolicy.py:create_obsolete_list:665] (MainThread) Obsolete item: rwaction LIST ENDS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,879  - DEBUG - [nitrointerface.py:delete_ns_config:4297] (MainThread) delete policy rewrite action k8s_rwaction_crd_proxyprotocol0<ns_namespace>_0_namespace_139420825 added to bulk config
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,910  - DEBUG - [rewritepolicy.py:create_obsolete_list:662] (MainThread) Obsolete Item: rwaudmesaction LIST BEGINS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,910  - DEBUG - [rewritepolicy.py:create_obsolete_list:665] (MainThread) Obsolete item: rwaudmesaction LIST ENDS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,943  - DEBUG - [rewritepolicy.py:create_obsolete_list:662] (MainThread) Obsolete Item: stringmap LIST BEGINS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:11,943  - DEBUG - [rewritepolicy.py:create_obsolete_list:665] (MainThread) Obsolete item: stringmap LIST ENDS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:12,085  - DEBUG - [rewritepolicy.py:create_obsolete_list:662] (MainThread) Obsolete Item: patset LIST BEGINS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:12,085  - DEBUG - [rewritepolicy.py:create_obsolete_list:665] (MainThread) Obsolete item: patset LIST ENDS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:12,116  - DEBUG - [rewritepolicy.py:create_obsolete_list:662] (MainThread) Obsolete Item: dataset LIST BEGINS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:12,116  - DEBUG - [rewritepolicy.py:create_obsolete_list:665] (MainThread) Obsolete item: dataset LIST ENDS
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:12,116  - DEBUG - [NetscalerConfigEngine.py:configure:31] (MainThread) clubconfig configure: DELETE : Removing obsolete binding lb vserver <ns_namespace>-netscaler-internal-443_namespace_443_<ns_namespace>-namespace-nginx-ingress-internal-controller_namespace_443_svc with rewrite policy k8s_rwpolicy_crd_proxyprotocol0<ns_namespace>_0_namespace_139420825
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:12,150  - DEBUG - [NetscalerConfigEngine.py:configure:31] (MainThread) clubconfig configure: DELETE : Removing obsolete binding lb vserver <ns_namespace>-netscaler-internal-80_namespace_80_<ns_namespace>-namespace-nginx-ingress-internal-controller_namespace_80_svc with rewrite policy k8s_rwpolicy_crd_proxyprotocol0<ns_namespace>_0_namespace_139420825
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:12,186  - DEBUG - [NetscalerConfigEngine.py:configure:31] (MainThread) clubconfig configure: DELETE : Removing obsolete rwpolicy k8s_rwpolicy_crd_proxyprotocol0<ns_namespace>_0_namespace_139420825
[netscaler-5bbbcd4c46-rcx7d] 2020-06-12 16:55:12,219  - DEBUG - [NetscalerConfigEngine.py:configure:31] (MainThread) clubconfig configure: DELETE : Removing obsolete rwaction k8s_rwaction_crd_proxyprotocol0<ns_namespace>_0_namespace_139420825
```

Seems to be solved in *1.7.46*. But now rewritePolicy only created for one servicePort(80), service has 80/443 with netscaler ingress for each port

**Comment by RaveendraHolla** [comment]
@rongo89 I am looking in to it.

**Comment by rongo89** [comment]
sure.

the first ingress for port 443
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    ingress.citrix.com/insecure-port: "443"
    ingress.citrix.com/insecure-service-type: tcp
    kubernetes.io/ingress.class: Citrix
  name: netscaler-internal-443
  namespace: namespace
spec:
  backend:
    serviceName: nginx-ingress-internal-controller
    servicePort: 443
```

the second ingress for port 80
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    ingress.citrix.com/insecure-service-type: tcp
    kubernetes.io/ingress.class: Citrix
  name: netscaler-internal-80
  namespace: namespace
spec:
  backend:
    serviceName: nginx-ingress-internal-controller
    servicePort: 80
```

the policy I want to have
```
apiVersion: citrix.com/v1
  kind: rewritepolicy
  metadata:
    name: proxypolicy
    namespace: namespace
  spec:
    rewrite-policies:
    - rewrite-policy:
        comment: Inject proxy protocol header v1
        direction: REQUEST
        modify-expression: '"PROXY TCP4 "+client.ip.src +" "+client.ip.dst +" "+client.tcp.srcport
          +" "+client.tcp.dstport +"\r\n"'
        operation: insert_before
        rewrite-criteria: "TRUE"
        target: client.TCP.PAYLOAD(1)
      servicenames:
      - nginx-ingress-internal-controller
  status:
    state: Success
    status_message: CRD Activated
```

The same happens if create LB via service with type loadbalancer. Only policy for one port created

**Comment by RaveendraHolla** [comment]
@rongo89 I can see the issue in my testbed. Analysing now. Will update soon.

**Comment by RaveendraHolla** [comment]
@rongo89 Got the root cause. Will push the fix in the next update. Work-around is to delete ingress and add again.

**Comment by rongo89** [comment]
thank you very much for the fast response, much appreciated @RaveendraHolla

**Comment by RaveendraHolla** [comment]
@rongo89 Fix now is available on CIC version 1.8.28.. Can you please verify this, so that I can close this Jira ticket.
https://github.com/citrix/citrix-k8s-ingress-controller/releases/tag/1.8.28

**Comment by rongo89** [comment]
problem solved
